packages/rollup-contracts/contracts/state-tree/FullPatriciaTree.sol

```diff
@@ -0,0 +1,380 @@
+            currentEdge.node = keccak256(abi.encode(edgeHashes[0], edgeHashes[1]));
+        }
+        // no more branching, so the remaining label is the root edge's label
+        currentEdge.label = remaining;
+        require(rootHash == edgeHash(currentEdge), 'Edge inclusion proof verification failed: root hashes do not match.');
+    }
+
+    function verifyProof(
+        bytes32 rootHash,
+        bytes32 key,
+        bytes memory value,
+        uint branchMask,
+        bytes32[] memory siblings
+    ) public pure {
+        // The edge above a leaf commits to the leaf value (i.e. what was actually set) 
+        bytes32 edgeCommittment = keccak256(value);
+        // The full "label" for a leaf node is the entirety of the key.
+        D.Label memory fullLabel = D.Label(key, 256);
+
+        verifyEdgeInclusionProof(
+            rootHash,
+            edgeCommittment,
+            fullLabel,
+            branchMask,
+            siblings
+        );
+    }
+
+    function verifyNonInclusionProof(
+        bytes32 rootHash,
+        bytes32 key,
+        bytes32 conflictingEdgeFullLabelData,
+        uint conflictingEdgeFullLabelLength,
+        bytes32 conflictingEdgeCommitment,
+        uint branchMask,
+        bytes32[] memory siblings
+    ) public pure {
+        // first, verify there is a conflict between the key and given edge
+        require(conflictingEdgeFullLabelLength <= 256, 'invalid label specified');
+        D.Label memory fullConflictingEdgeLabel = D.Label(conflictingEdgeFullLabelData, conflictingEdgeFullLabelLength);
+        D.Label memory fullLeafLabel = D.Label(key, 256);
+        uint indexOfConflict = Utils.commonPrefix(fullConflictingEdgeLabel, fullLeafLabel);
+        uint doesBranchAtConflict = branchMask & (1 << 255 - indexOfConflict);
+        
+        require(doesBranchAtConflict == 0, 'The provided conflicting edge is not actually conflicting.');
+        verifyEdgeInclusionProof(
+            rootHash,
+            conflictingEdgeCommitment,
+            fullConflictingEdgeLabel,
+            branchMask,
+            siblings
+        );
+    }
+
+    function insert(Tree storage tree, bytes32 key, bytes memory value) internal {
+        D.Label memory k = D.Label(key, 256);
+        bytes32 valueHash = keccak256(value);
+        tree.values[valueHash] = value;
+        // keys.push(key);
+        D.Edge memory e;
+        if (tree.rootEdge.node == 0 && tree.rootEdge.label.length == 0)
+        {
+            // Empty Trie
+            e.label = k;
+            e.node = valueHash;
+        }
+        else
+        {
+            e = _insertAtEdge(tree, tree.rootEdge, k, valueHash);
+        }
+        tree.root = edgeHash(e);
+        tree.rootEdge = e;
+    }
+
+    function _insertAtNode(
+        Tree storage tree, 
+        bytes32 nodeHash, 
+        D.Label memory key, 
+        bytes32 value
+    ) private returns (bytes32) {
+        require(key.length > 1, "Bad key");
```

Comment:
Would say why this key is bad. 

I've been confused looking at this PR since the names and the types in many cases don't indicate what is actually expected or represented. For instance, this accepts a key object, which is meant to represent a path, but it actually expects it to be a boolean. Code here and in many other places could be much clearer for a reviewer, contributor, auditor, etc. if this was indicated either by type or by name.